File: ios/MenuMakerUITests/PageObjects/SellerAnalyticsPage.swift
//
//  SellerAnalyticsPage.swift
//  MenuMakerUITests
//
//  Page Object for Seller Analytics/Dashboard Screen
//

import XCTest

struct SellerAnalyticsPage {
    let app: XCUIApplication

    // MARK: - Elements

    var totalSalesLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '₹' AND (label CONTAINS[c] 'total' OR label CONTAINS[c] 'sales')")).firstMatch
    }

    var totalOrdersLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'order' AND label MATCHES '.*\\\\d+.*'")).firstMatch
    }

    var totalRevenueLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '₹' AND label CONTAINS[c] 'revenue'")).firstMatch
    }

    var averageOrderValueLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '₹' AND label CONTAINS[c] 'average'")).firstMatch
    }

    var todayTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'today'")).firstMatch
    }

    var weekTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'week'")).firstMatch
    }

    var monthTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'month'")).firstMatch
    }

    var customRangeTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'custom'")).firstMatch
    }

    var salesChart: XCUIElement {
        app.otherElements.matching(identifier: "SalesChart").firstMatch
    }

    var popularItemsList: XCUIElementQuery {
        app.scrollViews.otherElements.matching(identifier: "PopularItem")
    }

    var firstPopularItem: XCUIElement {
        popularItemsList.firstMatch
    }

    var topSellingSection: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'top selling' OR label CONTAINS[c] 'popular'")).firstMatch
    }

    var peakHoursSection: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'peak' OR label CONTAINS[c] 'busy'")).firstMatch
    }

    var customerInsightsSection: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'customer' OR label CONTAINS[c] 'insights'")).firstMatch
    }

    var newCustomersLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'new customer' AND label MATCHES '.*\\\\d+.*'")).firstMatch
    }

    var repeatCustomersLabel: XCUIElement {
        app.staticTexs.matching(NSPredicate(format: "label CONTAINS[c] 'repeat' AND label MATCHES '.*\\\\d+.*'")).firstMatch
    }

    var averageRatingLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '★' OR (label CONTAINS[c] 'rating' AND label MATCHES '.*\\\\d+\\\\.\\\\d+.*')")).firstMatch
    }

    var totalReviewsLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'review' AND label MATCHES '.*\\\\d+.*'")).firstMatch
    }

    var exportButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'export' OR label CONTAINS[c] 'download'")).firstMatch
    }

    var refreshButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'refresh'")).firstMatch
    }

    var filterButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'filter'")).firstMatch
    }

    var pendingPayoutsLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '₹' AND label CONTAINS[c] 'pending'")).firstMatch
    }

    var completedPayoutsLabel: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '₹' AND label CONTAINS[c] 'completed'")).firstMatch
    }

    var dashboardCards: XCUIElementQuery {
        app.otherElements.matching(identifier: "DashboardCard")
    }

    var emptyStateMessage: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'no data' OR label CONTAINS[c] 'no sales'")).firstMatch
    }

    // MARK: - Actions

    @discardableResult
    func switchToTodayView() -> SellerAnalyticsPage {
        if todayTab.waitForExistence(timeout: 2) {
            todayTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToWeekView() -> SellerAnalyticsPage {
        if weekTab.waitForExistence(timeout: 2) {
            weekTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToMonthView() -> SellerAnalyticsPage {
        if monthTab.waitForExistence(timeout: 2) {
            monthTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToCustomRange() -> SellerAnalyticsPage {
        if customRangeTab.waitForExistence(timeout: 2) {
            customRangeTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func tapFirstPopularItem() -> SellerAnalyticsPage {
        if firstPopularItem.waitForExistence(timeout: 2) {
            firstPopularItem.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func tapExport() -> SellerAnalyticsPage {
        if exportButton.waitForExistence(timeout: 2) {
            exportButton.tap()
            sleep(1)

            // Handle export options if sheet appears
            let csvOption = app.buttons["CSV"]
            let pdfOption = app.buttons["PDF"]

            if csvOption.waitForExistence(timeout: 1) {
                csvOption.tap()
            } else if pdfOption.waitForExistence(timeout: 1) {
                pdfOption.tap()
            }
        }
        return self
    }

    @discardableResult
    func refreshDashboard() -> SellerAnalyticsPage {
        if refreshButton.waitForExistence(timeout: 1) {
            refreshButton.tap()
        } else {
            // Try pull to refresh
            let scrollView = app.scrollViews.firstMatch
            if scrollView.exists {
                let start = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.2))
                let end = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.8))
                start.press(forDuration: 0, thenDragTo: end)
            }
        }
        sleep(2)
        return self
    }

    @discardableResult
    func scrollToSection(_ section: AnalyticsSection) -> SellerAnalyticsPage {
        let scrollView = app.scrollViews.firstMatch
        if scrollView.exists {
            switch section {
            case .popularItems:
                scrollView.swipeUp()
            case .peakHours:
                scrollView.swipeUp()
                scrollView.swipeUp()
            case .customerInsights:
                scrollView.swipeUp()
                scrollView.swipeUp()
                scrollView.swipeUp()
            case .payouts:
                scrollView.swipeUp()
                scrollView.swipeUp()
                scrollView.swipeUp()
                scrollView.swipeUp()
            }
        }
        return self
    }

    @discardableResult
    func tapFilter() -> SellerAnalyticsPage {
        if filterButton.waitForExistence(timeout: 2) {
            filterButton.tap()
            sleep(1)
        }
        return self
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> SellerAnalyticsPage {
        XCTAssertTrue(totalSalesLabel.waitForExistence(timeout: timeout) ||
                     totalOrdersLabel.waitForExistence(timeout: timeout) ||
                     salesChart.waitForExistence(timeout: timeout) ||
                     dashboardCards.count > 0,
                     "Analytics dashboard should be displayed")
        return self
    }

    @discardableResult
    func assertSalesMetricsDisplayed() -> SellerAnalyticsPage {
        XCTAssertTrue(totalSalesLabel.exists ||
                     totalOrdersLabel.exists ||
                     totalRevenueLabel.exists,
                     "Sales metrics should be displayed")
        return self
    }

    @discardableResult
    func assertChartDisplayed() -> SellerAnalyticsPage {
        XCTAssertTrue(salesChart.exists, "Sales chart should be displayed")
        return self
    }

    @discardableResult
    func assertTimePeriodsDisplayed() -> SellerAnalyticsPage {
        XCTAssertTrue(todayTab.exists ||
                     weekTab.exists ||
                     monthTab.exists,
                     "Time period tabs should be displayed")
        return self
    }

    @discardableResult
    func assertPopularItemsDisplayed() -> SellerAnalyticsPage {
        XCTAssertTrue(firstPopularItem.exists ||
                     topSellingSection.exists,
                     "Popular items section should be displayed")
        return self
    }

    @discardableResult
    func assertCustomerInsightsDisplayed() -> SellerAnalyticsPage {
        XCTAssertTrue(newCustomersLabel.exists ||
                     repeatCustomersLabel.exists ||
                     customerInsightsSection.exists,
                     "Customer insights should be displayed")
        return self
    }

    @discardableResult
    func assertRatingDisplayed() -> SellerAnalyticsPage {
        XCTAssertTrue(averageRatingLabel.exists ||
                     totalReviewsLabel.exists,
                     "Rating information should be displayed")
        return self
    }

    @discardableResult
    func assertPayoutsDisplayed() -> SellerAnalyticsPage {
        XCTAssertTrue(pendingPayoutsLabel.exists ||
                     completedPayoutsLabel.exists,
                     "Payout information should be displayed")
        return self
    }

    @discardableResult
    func assertExportButtonVisible() -> SellerAnalyticsPage {
        XCTAssertTrue(exportButton.exists, "Export button should be visible")
        return self
    }

    @discardableResult
    func assertEmptyState() -> SellerAnalyticsPage {
        XCTAssertTrue(emptyStateMessage.exists, "Empty state should be displayed")
        return self
    }

    @discardableResult
    func assertDataUpdated() -> SellerAnalyticsPage {
        // After refresh, data should be displayed
        XCTAssertTrue(totalSalesLabel.exists ||
                     totalOrdersLabel.exists,
                     "Data should be updated")
        return self
    }

    // MARK: - Types

    enum AnalyticsSection {
        case popularItems
        case peakHours
        case customerInsights
        case payouts
    }
}

File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/SellerAnalyticsPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Seller Analytics/Dashboard Screen
 * Provides fluent API for analytics interactions
 */
class SellerAnalyticsPage(private val composeTestRule: ComposeTestRule) {

    // Metric labels
    private val totalSalesLabel = composeTestRule.onNode(
        hasText("₹", substring = true) and
        (hasText("total", substring = true, ignoreCase = true) or hasText("sales", substring = true, ignoreCase = true))
    )
    private val totalOrdersLabel = composeTestRule.onNodeWithText("order", substring = true, ignoreCase = true)
    private val totalRevenueLabel = composeTestRule.onNode(
        hasText("₹", substring = true) and hasText("revenue", substring = true, ignoreCase = true)
    )
    private val averageOrderValueLabel = composeTestRule.onNode(
        hasText("₹", substring = true) and hasText("average", substring = true, ignoreCase = true)
    )

    // Time period tabs
    private val todayTab = composeTestRule.onNodeWithText("today", substring = true, ignoreCase = true)
    private val weekTab = composeTestRule.onNodeWithText("week", substring = true, ignoreCase = true)
    private val monthTab = composeTestRule.onNodeWithText("month", substring = true, ignoreCase = true)
    private val customRangeTab = composeTestRule.onNodeWithText("custom", substring = true, ignoreCase = true)

    // Chart and items
    private val salesChart = composeTestRule.onNodeWithTag("SalesChart")
    private val popularItemsList = composeTestRule.onAllNodesWithTag("PopularItem")
    private val topSellingSection = composeTestRule.onNode(
        hasText("top selling", substring = true, ignoreCase = true) or
        hasText("popular", substring = true, ignoreCase = true)
    )

    // Customer insights
    private val peakHoursSection = composeTestRule.onNode(
        hasText("peak", substring = true, ignoreCase = true) or
        hasText("busy", substring = true, ignoreCase = true)
    )
    private val customerInsightsSection = composeTestRule.onNode(
        hasText("customer", substring = true, ignoreCase = true) or
        hasText("insights", substring = true, ignoreCase = true)
    )
    private val newCustomersLabel = composeTestRule.onNodeWithText("new customer", substring = true, ignoreCase = true)
    private val repeatCustomersLabel = composeTestRule.onNodeWithText("repeat", substring = true, ignoreCase = true)

    // Rating and reviews
    private val averageRatingLabel = composeTestRule.onNode(
        hasText("★", substring = true) or
        hasText("rating", substring = true, ignoreCase = true)
    )
    private val totalReviewsLabel = composeTestRule.onNodeWithText("review", substring = true, ignoreCase = true)

    // Buttons
    private val exportButton = composeTestRule.onNode(
        hasText("export", substring = true, ignoreCase = true) or
        hasText("download", substring = true, ignoreCase = true)
    )
    private val refreshButton = composeTestRule.onNodeWithContentDescription("refresh", substring = true, ignoreCase = true)
    private val filterButton = composeTestRule.onNodeWithText("filter", substring = true, ignoreCase = true)

    // Payouts
    private val pendingPayoutsLabel = composeTestRule.onNode(
        hasText("₹", substring = true) and hasText("pending", substring = true, ignoreCase = true)
    )
    private val completedPayoutsLabel = composeTestRule.onNode(
        hasText("₹", substring = true) and hasText("completed", substring = true, ignoreCase = true)
    )

    // Dashboard cards
    private val dashboardCards = composeTestRule.onAllNodesWithTag("DashboardCard")
    private val emptyStateMessage = composeTestRule.onNode(
        hasText("no data", substring = true, ignoreCase = true) or
        hasText("no sales", substring = true, ignoreCase = true)
    )

    // Actions
    fun switchToTodayView(): SellerAnalyticsPage {
        todayTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun switchToWeekView(): SellerAnalyticsPage {
        weekTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun switchToMonthView(): SellerAnalyticsPage {
        monthTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun switchToCustomRange(): SellerAnalyticsPage {
        customRangeTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun tapFirstPopularItem(): SellerAnalyticsPage {
        if (popularItemsList.fetchSemanticsNodes().isNotEmpty()) {
            popularItemsList[0].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun tapExport(): SellerAnalyticsPage {
        exportButton.performClick()
        Thread.sleep(1000)

        // Handle export options if dialog appears
        val csvOption = composeTestRule.onNodeWithText("CSV")
        val pdfOption = composeTestRule.onNodeWithText("PDF")

        try {
            csvOption.performClick()
        } catch (e: AssertionError) {
            try {
                pdfOption.performClick()
            } catch (e: AssertionError) {
                // No export options found
            }
        }
        return this
    }

    fun refreshDashboard(): SellerAnalyticsPage {
        refreshButton.performClick()
        Thread.sleep(2000)
        return this
    }

    fun tapFilter(): SellerAnalyticsPage {
        filterButton.performClick()
        Thread.sleep(1000)
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): SellerAnalyticsPage {
        composeTestRule.waitUntil(timeoutMillis = 2000) {
            try { totalSalesLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { totalOrdersLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { salesChart.assertExists(); true } catch (e: AssertionError) { false } ||
            dashboardCards.fetchSemanticsNodes().isNotEmpty()
        }
        return this
    }

    fun assertSalesMetricsDisplayed(): SellerAnalyticsPage {
        assert(
            try { totalSalesLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { totalOrdersLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { totalRevenueLabel.assertExists(); true } catch (e: AssertionError) { false }
        ) {
            "Sales metrics should be displayed"
        }
        return this
    }

    fun assertChartDisplayed(): SellerAnalyticsPage {
        salesChart.assertExists()
        return this
    }

    fun assertTimePeriodsDisplayed(): SellerAnalyticsPage {
        todayTab.assertExists()
        return this
    }

    fun assertPopularItemsDisplayed(): SellerAnalyticsPage {
        assert(
            popularItemsList.fetchSemanticsNodes().isNotEmpty() ||
            try { topSellingSection.assertExists(); true } catch (e: AssertionError) { false }
        ) {
            "Popular items section should be displayed"
        }
        return this
    }

    fun assertCustomerInsightsDisplayed(): SellerAnalyticsPage {
        assert(
            try { newCustomersLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { repeatCustomersLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { customerInsightsSection.assertExists(); true } catch (e: AssertionError) { false }
        ) {
            "Customer insights should be displayed"
        }
        return this
    }

    fun assertRatingDisplayed(): SellerAnalyticsPage {
        assert(
            try { averageRatingLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { totalReviewsLabel.assertExists(); true } catch (e: AssertionError) { false }
        ) {
            "Rating information should be displayed"
        }
        return this
    }

    fun assertPayoutsDisplayed(): SellerAnalyticsPage {
        assert(
            try { pendingPayoutsLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { completedPayoutsLabel.assertExists(); true } catch (e: AssertionError) { false }
        ) {
            "Payout information should be displayed"
        }
        return this
    }

    fun assertExportButtonVisible(): SellerAnalyticsPage {
        exportButton.assertExists()
        return this
    }

    fun assertEmptyState(): SellerAnalyticsPage {
        emptyStateMessage.assertExists()
        return this
    }

    fun assertDataUpdated(): SellerAnalyticsPage {
        assert(
            try { totalSalesLabel.assertExists(); true } catch (e: AssertionError) { false } ||
            try { totalOrdersLabel.assertExists(); true } catch (e: AssertionError) { false }
        ) {
            "Data should be updated"
        }
        return this
    }

    enum class AnalyticsSection {
        POPULAR_ITEMS,
        PEAK_HOURS,
        CUSTOMER_INSIGHTS,
        PAYOUTS
    }
}
